lib/jxl/render_pipeline/stage_write.cc

```diff
@@ -356,20 +322,193 @@ class WriteToPixelCallbackStage : public RenderPipelineStage {
             undo_orientation == Orientation::kRotate270 ||
             undo_orientation == Orientation::kAntiTranspose);
   }
+
+  void UnpremulAlpha(size_t thread_id, size_t len,
+                     const float** line_buffers) const {
+    const HWY_FULL(float) d;
+    auto one = Set(d, 1.0f);
+    float* temp_in[4];
+    for (size_t c = 0; c < num_channels_; ++c) {
+      size_t tix = thread_id * num_channels_ + c;
+      temp_in[c] = reinterpret_cast<float*>(temp_in_[tix].get());
+      memcpy(temp_in[c], line_buffers[c], sizeof(float) * len);
+      UnpoisonPadding(d, len, temp_in[c]);
+    }
+    auto small_alpha = Set(d, kSmallAlpha);
+    for (size_t ix = 0; ix < len; ix += Lanes(d)) {
+      auto alpha = LoadU(d, temp_in[num_color_] + ix);
+      auto mul = Div(one, Max(small_alpha, alpha));
+      for (size_t c = 0; c < num_color_; ++c) {
+        auto val = LoadU(d, temp_in[c] + ix);
+        StoreU(Mul(val, mul), d, temp_in[c] + ix);
+      }
+    }
+    for (size_t c = 0; c < num_channels_; ++c) {
+      line_buffers[c] = temp_in[c];
+    }
+  }
+
+  void FlipX(size_t thread_id, size_t len, size_t* xstart,
+             const float** line_buffers) const {
+    float* temp_in[4];
+    for (size_t c = 0; c < num_channels_; ++c) {
+      size_t tix = thread_id * num_channels_ + c;
+      temp_in[c] = reinterpret_cast<float*>(temp_in_[tix].get());
+      if (temp_in[c] != line_buffers[c]) {
+        memcpy(temp_in[c], line_buffers[c], sizeof(float) * len);
+      }
+    }
+    size_t last = (len - 1u);
+    size_t num = (len / 2);
+    for (size_t i = 0; i < num; ++i) {
+      for (size_t c = 0; c < num_channels_; ++c) {
+        std::swap(temp_in[c][i], temp_in[c][last - i]);
+      }
+    }
+    for (size_t c = 0; c < num_channels_; ++c) {
+      line_buffers[c] = temp_in[c];
+    }
+    *xstart = width_ - *xstart - len;
+  }
+
   template <typename T>
-  void WriteToCallback(size_t thread_id, size_t ypos, size_t xstart,
-                       size_t xlen, T* output) const {
+  void StoreUnsignedRow(const float* input[4], size_t len, T* output) const {
```

Comment:
Does this work with 16 or 8 bit types?

Reply:
It works with uint8_t and uint16_t.

Reply:
Uh, OK, I thought DemoteTo didn't... Oh well, better this way :)